_posts/2023-08-02-turbocharging-rust-code-verification.md

```diff
@@ -0,0 +1,446 @@
+Kani can now perform the MIR-to-GOTO code generation and GOTO binary export 4x faster than before.
+
+The table below reports total time and memory consumption when running `kani --tests --only-codegen` with `kani-0.33.3` for three crates of the `s2n-quic` project, with JSON symbol table export versus GOTO binary export.
+
+| Crate                | User time (JSON) | User time (GOTO bin.) | Peak Mem. (JSON) | Peak Mem. (GOTO bin.) |
+| -------------------  | ---------------: | --------------------: | ---------------: | --------------------: |
+| `s2n-quic-core`      |             198s |              **111s** |        **669Mb** |                 815Mb |
+| `s2n-quic-platform`  |              99s |               **91s** |        **449Mb** |                 450Mb |
+| `s2n-quic-xdp`       |              92s |               **95s** |        **450Mb** |                 450Mb |
+
+
+For `s2n-quic-core`, which contains 37 verification harnesses, we observe a reduction of 45% of *User time* from 199s down to 111s, and a 20% memory consumption increase from ~670Mb to ~815Mb.
+For the two other crates which contain respectively 3 and 1 harnesses, the *User time* is dominated by Rust to MIR compilation and the total gains with GOTO binary export are lower
+at roughly 10% and even degrade to negative 4%, albeit for a virtually non memory consumption increase.
+
+Looking in more detail at GOTO code generation and GOTO binary export time for individual harnesses, excluding the Rust-to-MIR compilation time,
+we see that GOTO binary export is ~9x faster than JSON export, and that the GOTO code generation and export step is now ~4x faster (see the _Detailed table_ below). The savings seem small but accumulate to eventually make a noticeable difference.
+
+GOTO binary export is now the default export mode, because it is faster when there are multiple verification harnesses in a crate, which is more often the case in practice.
+The memory consumption can be sometimes higher than with JSON export, but it can also sometimes be lower depending on how much opportunity for sharing of identical subtrees the crate offers (on large crates like `std` or `proc_macro` we've observed up to 2x reduction in memory usage). Should you ever need it, JSON export can still be activated with the command line switch `--write-json-symbtab`.
+
+
+<details>
+<summary>
+Detailed GOTO codegen and export times table
+</summary>
+<div markdown=1>
+
+| Harness                                                     | GOTO codegen time (s) | JSON export (s) | GOTO-binary export (s) | export speedup | codegen + export speedup |
+| ----------------------------------------------------------- | --------------------: | --------------: | ---------------------: | -------------: | -----------------------: |
+| `ct::tests::ct_ge...`                                       |                  0,32 |            1,48 |               **0,17** |           8,70 |                     3,67 |
```

Comment:
In the US it's more common to report decimals using `.` instead of `,`

Reply:
fixed